### PR TITLE
Fixes to enable lms to use gde

### LINKED ISSUE #7: Port a fork of tensorflow-large-model-support to GraphDef Editor
Create an experimental fork of the [TensorFlow Large Model Support](https://github.com/IBM/tensorflow-large-model-support) project. Port the LMS library to use GraphDef Editor in place of `contrib.graph_editor`.

### PR DESCRIPTION
**Pull request by BryanCutler**
Fixes include

* Check Graph collection type for tensor types
* Add methods to Tensor `add_to_collection` and `collection_names`
* Add utility to create an identity op
* Fixed bug in `add_control_inputs`
* Fixed bug in subgraph when getting view with op and graph kwarg

Fixes #7

### CODE REVIEW

**Comment by BryanCutler** [comment]
Still a bit of a WIP, need to add unit tests but it's running LMS mnist example!

**Comment by BryanCutler** [comment]
@frreiss , running the `cmm_mnist_lms` example chokes on
```
File "/home/bryan/miniconda2/envs/gde-dev/lib/python3.6/site-packages/graph_def_editor/graph.py", line 1178, in _make_collection_defs
    first_item_type.__name__))
NotImplementedError: Can't serialize item '<tensorflow.python.estimator.inputs.queues.feeding_queue_runner._FeedingQueueRunner object at 0x7f92ce397dd8>' in collection 'queue_runners' because it is a '_FeedingQueueRunner'.
```
Is that type of collection something that needs to be in GDE?

File: tests/util_test.py [hunk 1]
```diff
@@ -180,6 +180,23 @@ def test_placeholder(self):
...
+    g = gde.Graph(tf_g)
+    i2 = gde.util.make_identity(g, "identity_gde", g.get_tensor_by_name(c.name))
+
+    with g.to_tf_graph().as_default():
+      with tf.Session() as sess:
+        result1 = sess.run(i1.name)
+        result2 = sess.run(i2.name)
+      # TODO: test fails
+      print(tf.get_default_graph().as_graph_def())
+    self.assertEqual(result1, result2)
```
Comment (BryanCutler) [inline tests/util_test.py]: @frreiss this fails because the created identity op return `None` for result2. The graph def output looks the same, and I'm not sure what is wrong, could you take a look at `util.make_identity` when you have a sec? thanks!

File: graph_def_editor/graph.py [hunk 2]
```diff
@@ -242,7 +242,11 @@ def add_collection_from_collection_def(
     # The collection is stored in exactly one of five different formats.
     if collection.HasField("node_list"):
       for node_name in collection_def.value.node_list.value:
-        n = self.get_node_by_name(node_name)
+        # Check if node name is a Tensor type
+        if node_name.rfind(':') > -1:
```
[the remark this answers is not in the document]
Comment (frreiss) [reply graph_def_editor/graph.py]: Unfortunately, looking for the ":" is the only way to know if the string is a tensor. Or at least I haven't found anything better. The code that creates those `CollectionDef` protocol buffer messages is at https://github.com/tensorflow/tensorflow/blob/6179f7d477260e5628fc661b9ff31bc4fd770d54/tensorflow/python/framework/meta_graph.py#L377 and is not terribly helpful. Reading between the lines, it looks like the serialization code treats collections of Python `string` objects as "node names", regardless fo what those strings are supposed to represent. Whoever was writing the code to store collections of tensors, as in node outputs -- as opposed to tensors as in multidimensional arrays -- didn't want to bother with adding a new field to the protobuf definition or changing the code in `meta_graph.py`.

TL;DR write something that works for the examples you have on hand and add some warning comments around it.

File: graph_def_editor/util.py [hunk 3]
```diff
@@ -730,3 +730,25 @@ def make_placeholder(g: 'graph.Graph', name: str, dtype: tf.DType,
...
+    uniquify_name: if True, generate unique names by appending a numeric
+      suffix in the event of a name collision. Otherwise name collisions
+      result in an error.
+
+  Returns `gde.Node` object representing the new node.
+  """
+  ret = g.add_node(name, "Identity", uniquify_name=uniquify_name)
+  ret.set_inputs([input])
+  ret.set_outputs_from_pairs([(input.dtype, input.shape)])
+  ret.add_attr("T", input.dtype)
```
Comment (frreiss) [inline graph_def_editor/util.py]: Can you also copy the input shape here (if present)?

File: tests/graph_test.py [hunk 4]
```diff
@@ -120,6 +120,99 @@ def test_export_saved_model_with_var(self):
...
+    after_tf_g = tf.Graph()
+    with after_tf_g.as_default():
+      with tf.Session() as sess:
+        tf.saved_model.load(sess, [tf.saved_model.tag_constants.SERVING],
+                            second_model_dir)
+
+    # Checking collections loaded back from gde savedmodel
+    for name in expected_collections:
+      self.assertIn(name, after_tf_g.collections)
+
```
Comment (frreiss) [inline tests/graph_test.py]: You should probably repeat the more specific checks up at line 193 here to ensure that the collection contents got properly reconstituted by `tf.saved_model.load()`

File: tests/util_test.py (hunk 1 above)
Comment (frreiss) [reply tests/util_test.py]: `i2` is a Node, not a tensor. So line 195 should be:
```
result2 = sess.run(i2.output(0).name)
```

**Comment by frreiss** [comment]
> @frreiss , running the `cmm_mnist_lms` example chokes on
> 
> ```
> File "/home/bryan/miniconda2/envs/gde-dev/lib/python3.6/site-packages/graph_def_editor/graph.py", line 1178, in _make_collection_defs
>     first_item_type.__name__))
> NotImplementedError: Can't serialize item '<tensorflow.python.estimator.inputs.queues.feeding_queue_runner._FeedingQueueRunner object at 0x7f92ce397dd8>' in collection 'queue_runners' because it is a '_FeedingQueueRunner'.
> ```
> Is that type of collection something that needs to be in GDE?

Yeah, anything that is serialized as part of the GraphDef/MetaGraphDef needs to at least be passed through by GDE. Otherwise the graph won't work properly for some use cases after it's been round-tripped through our library.

File: graph_def_editor/util.py (hunk 3 above)
Comment (BryanCutler) [reply graph_def_editor/util.py]: the shape is in the output, you mean add it to attr? I didn't see that as part of `tf.identity.as_graph_def()`

File: graph_def_editor/graph.py [hunk 5]
```diff
@@ -776,6 +783,9 @@ def _build_collection_name_to_type(self):
     for n in self.nodes:
       for name in n.collection_names:
         node_collection_names.add(name)
+    for t in self.tensors:
```
[the remark this answers is not in the document]
Comment (BryanCutler) [reply graph_def_editor/graph.py]: How about just check if the `node_collection` set is empty before adding in tensors?

**Comment by BryanCutler** [comment]
Thanks for the review @frreiss! I think I addressed all the feedback except for a question here https://github.com/CODAIT/graph_def_editor/pull/17#discussion_r251176268
